grayskull/strategy/py_build.py

```diff
@@ -0,0 +1,144 @@
+for inlinining into the first-party project source tree.
+"""
+
+import logging
+import tempfile
+from importlib.metadata import PathDistribution
+from pathlib import Path
+
+import build
+from conda.exceptions import InvalidMatchSpec
+from conda.models.match_spec import MatchSpec
+from packaging.requirements import Requirement
+from souschef.recipe import Recipe
+
+from grayskull.config import Configuration
+from grayskull.strategy.abstract_strategy import AbstractStrategy
+from grayskull.strategy.pypi import compose_test_section
+
+log = logging.getLogger(__name__)
+
+
+class PyBuild(AbstractStrategy):
+    @staticmethod
+    def fetch_data(recipe: Recipe, config: Configuration, sections=None):
+        project = build.ProjectBuilder(config.name)
+
+        recipe["source"]["path"] = "../"
+        # XXX relative to output. "git_url: ../" is also a good choice.
+
+        with tempfile.TemporaryDirectory(prefix="grayskull") as output:
+            build_system_requires = project.build_system_requires
+            requires_for_build = project.get_requires_for_build("wheel")
+            # If those are already installed, we can get the extras requirements
+            # without invoking pip e.g. setuptools_scm[toml]
+            print("Requires for build:", build_system_requires, requires_for_build)
+
+            # build the project's metadata "dist-info" directory
+            metadata_path = Path(project.metadata_path(output_directory=output))
+
+            distribution = PathDistribution(metadata_path)
+
+            # real distribution name not pathname
+            config.name = distribution.name  # see also _normalized_name
+            config.version = distribution.version
+
+            # grayskull thought the name was the path. correct that.
+            if recipe[0] == '#% set name = "." %}':  # XXX fragile
+                # recipe[0] = x does not work
+                recipe._yaml._yaml_get_pre_comment()[0].value = (
+                    f'#% set name = "{config.name}" %}}\n'
+                    f'#% set version = "{config.version}" %}}\n'
+                )
+
+                recipe["package"]["version"] = "{{ version }}"
+            elif config.name not in str(recipe[0]):
+                log.warning("Package name not found in first line of recipe")
+
+            metadata = distribution.metadata
+
+            requires_python = metadata["requires-python"]
+            if requires_python:
+                requires_python = f"python { requires_python }"
+            else:
+                requires_python = "python"
+
+            recipe["requirements"]["host"] = [requires_python] + sorted(
+                (*build_system_requires, *requires_for_build)
+            )
+
+            requirements = [Requirement(r) for r in distribution.requires]
+            active_requirements = [
+                str(r).rsplit(";", 1)[0]
+                for r in requirements
+                if not r.marker or r.marker.evaluate()
+            ]
```

Comment:
How do we include the marker as a YAML comment?

Reply:
These are evaluated in the environment grayskull runs in, and might not be correct for the end user.

Reply:
there is a property called ``.comment`` that you can use for that
a recipe/section object uses this mixin
https://github.com/marcelotrevisani/souschef/blob/92cee9159fb65fa035593f1a8b60a931413ab581/src/souschef/mixins.py#L14